File: web/src/types/ws.ts
export type WsTicketResponse = {
  ticket: string;
  expires_in: number;
};

export type WsEventMeta = {
  dedupe_key?: string | null;
  requires_refetch: string[];
  priority: "low" | "normal" | "high";
};

export type WsEventEnvelope = {
  id: string;
  topic: string;
  name: string;
  version: number;
  timestamp: string;
  payload: Record<string, unknown>;
  meta?: WsEventMeta | null;
};

export type WsReadyMessage = {
  type: "ready";
  connection_id: string;
  user_id: string;
  auto_topics: string[];
};

export type WsSubscribedMessage = {
  type: "subscribed";
  topics: string[];
  rejected: Array<{ topic: string; reason: string }>;
};

export type WsUnsubscribedMessage = {
  type: "unsubscribed";
  topics: string[];
  reason?: string;
};

export type WsEventMessage = {
  type: "event";
  event: WsEventEnvelope;
};

export type WsErrorMessage = {
  type: "error";
  code: string;
  message: string;
};

export type WsPongMessage = {
  type: "pong";
};

export type WsServerMessage =
  | WsReadyMessage
  | WsSubscribedMessage
  | WsUnsubscribedMessage
  | WsEventMessage
  | WsErrorMessage
  | WsPongMessage;

File: api/app/schemas/ws.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class WsTicketResponse(BaseModel):
    ticket: str
    expires_in: int


class WsRejectedTopic(BaseModel):
    topic: str
    reason: str


class WsEventMeta(BaseModel):
    dedupe_key: str | None = None
    requires_refetch: list[str] = Field(default_factory=list)
    priority: Literal["low", "normal", "high"] = "normal"


class WsEventEnvelope(BaseModel):
    id: str
    topic: str
    name: str
    version: int = 1
    timestamp: datetime
    payload: dict[str, Any] = Field(default_factory=dict)
    meta: WsEventMeta | None = None
